mqtt/mqtt-edgehub/src/auth/authorization/edgehub.rs

```diff
@@ -119,6 +122,64 @@ impl EdgeHubAuthorizer {
             .iter()
             .any(|allowed_topic| topic.starts_with(allowed_topic))
     }
+
+    fn get_on_behalf_of_id(topic: &str) -> Option<String> {
+        // topics without the new topic format cannot have on_behalf_of_ids
+        if !topic.starts_with("$iothub/clients") {
+            return None;
+        }
+        let topic_parts = topic.split('/').collect::<Vec<&str>>();
+        let device_id = topic_parts.get(2);
+        let module_id = match topic_parts.get(3) {
+            Some(s) => {
+                if *s == "modules" {
+                    topic_parts.get(4)
+                } else {
+                    None
+                }
+            }
+            None => None,
+        };
+        match module_id {
+            Some(mid) => match device_id {
+                Some(did) => Some(format!("{}/{}", did, mid)),
+                None => None,
+            },
+            None => match device_id {
+                Some(id) => Some((*id).to_string()),
+                None => None,
+            },
+        }
+    }
+
+    fn check_authorized_cache(&self, client_id: &ClientId, topic: &str) -> bool {
```

Comment:
I would restructure it like that. I believe it would be easier to follow the logic:
```rust
    fn check_authorized_cache(&self, client_id: &ClientId, topic: &str) -> bool {
        if let Some(on_behalf_of_id) = EdgeHubAuthorizer::get_on_behalf_of_id(topic) {
            if client_id.as_str() == on_behalf_of_id {
                self.service_identities_cache
                    .contains_key(client_id.as_str())
            } else {
                match self.service_identities_cache.get(&on_behalf_of_id) {
                    Some(service_identity) => {
                        match service_identity.auth_chain() {
                            Some(auth_chain) => {
                                // if auth_chain for on_behalf_of_id contains client_id, then on_behalf_of_id is a descendant of client_id
                                auth_chain.contains(client_id.as_str())
                            }
                            None => false,
                        }
                    }
                    None => false,
                }
            }
        } else {
            // If there is no on_behalf_of_id, we are dealing with a legacy topic
            // The client_id must still be in the identities cache
            self.service_identities_cache
                .contains_key(client_id.as_str())
        }
    }
```